Add quality flag filter documentation and examples (#201)
SolarArbiter/SolarArbiter.github.io: documentation/dashboard/working_with_data.md

```
Quality Flags: "USER FLAGGED"
Discard Before Resample: False
Resample Threshold Percentage: 50%
```
the observation data above is filtered and resampled into:
```
timestamp,value
2020-01-01 20:00:00+00:00,225.0
2020-01-01 21:00:00+00:00,130.0
```

#### Discarding data before resample

If we choose to discard data before resample, data points will be dropped
before the timeseries is resampled into a new interval.

Using the filter:
```
Quality Flags: "USER FLAGGED"
Discard Before Resample: True
Resample Threshold Percentage: 49%
```

the observation data above is filtered and resampled into: 
```
timestamp,value
2020-01-01 20:00:00+00:00,210.0
2020-01-01 21:00:00+00:00,130.0
```

In this case, the first interval is included in the resampled data because only
two of the four points in the first hour are flagged with `USER FLAGGED`. However,
the datapoints flagged with `USER FLAGGED` are removed before resampling is performed.
So the value of the first hour is calculated as the average of the points not flagged
with `USER FLAGGED` ((270 + 150) /2  = 210).

#### Filtering for multiple flags

When listing multiple flags in a single filter, the number of unique datapoints

> flagged with any of the listed quality flags is used when determining if the

Reviewer: ```suggestion
flagged with *any* of the listed quality flags determines if the
```

```
Quality Flags: "USER FLAGGED"
Discard Before Resample: True
Resample Threshold Percentage: 50%
```

the observation data above is filtered and resampled into: 
```
timestamp,value
2020-01-01 12:00:00+00:00,113.33
2020-01-01 13:00:00+00:00,155.0
```

In this case, the first interval is included in the resampled data because only
one point is flagged with `USER FLAGGED`, which does not exceed our 50% resample
threshold. However, the datapoint flagged with `USER FLAGGED` was dropped before
resampling occured which reduces the interval average from 235.0 to 113.33.

#### Filtering for multiple flags

When listing multiple flags in a single filter, the number of unique datapoints
flagged with *any* of the listed quality flags determines if the
resample threshold percentage is exceeded or not.

Using the filter: 
```
Quality Flags: "NIGHTTIME USER FLAGGED, "
Discard Before Resample: False
Resample Threshold Percentage: 50%
```

the observation data above is filtered and resampled into:
```
timestamp,value
2020-01-01 13:00:00+00:00,155.0
```
The first hour interval is dropped because three of the four timestamps in
the first hour are flagged with either `NIGHTTIME` or `USER FLAGGED`,

> exceeding the resample threshold.

Author: Added an example using these filters.

Resample Threshold Percentage: 50%
```

the observation data above is filtered and resampled into:
```
timestamp,value
2020-01-01 13:00:00+00:00,155.0
```
The first and second hour interval is dropped because 100% and 50% of the datapoints
in those intervals are flagged with `NIGHTTIME`, *meeting* or *exceeding* the
resample threshold.

If we want to include intervals where half of the data are
flagged with `NIGHTTIME`, we could adjust the resample threshold
percentage to 51%. The threshold must be *met or exceeded* for an interval
to be excluded.

Using the filter:
```
Quality Flags: "NIGHTTIME"
Discard Before Resample: False
Resample Threshold Percentage: 51%
```
the observation data above is filtered and resampled into:
```
timestamp,value
2020-01-01 12:00:00+00:00,235.0
2020-01-01 13:00:00+00:00,155.0
```

#### Discarding erroneous data

If we want to remove erroneous data and exclude it from the resampling
process, we can set discard before resample to True. 

Using the filter:
```
Quality Flags: "USER FLAGGED"
Discard Before Resample: True
Resample Threshold Percentage: 50%
```

the observation data above is filtered and resampled into: 
```
timestamp,value

> 2020-01-01 12:00:00+00:00,113.33

Reviewer: ```suggestion
2020-01-01 11:00:00+00:00,0.0
2020-01-01 12:00:00+00:00,113.33
```

2020-01-01 13:00:00+00:00,155.0
```

> In this case, the first interval is included in the resampled data because only
> one point is flagged with `USER FLAGGED`, which does not exceed our 50% resample
> threshold. However, the datapoint flagged with `USER FLAGGED` was dropped before
> resampling occured which reduces the interval average from 235.0 to 113.33.

Reviewer: ```suggestion
The datapoint flagged with `USER FLAGGED` was dropped before
resampling occurred, which reduces the 12:00 interval average from 235.0 to 113.33. The resampled interval is included in the resampled data because only
one point is flagged with `USER FLAGGED`, which does not exceed our 50% resample
threshold. 
```

#### Filtering for multiple flags in a single filter

When listing multiple flags in a single filter, the number of unique datapoints
flagged with *any* of the listed quality flags determines if the
resample threshold percentage is exceeded or not.

Using the filter: 
```

> Quality Flags: "NIGHTTIME, USER FLAGGED, "

Reviewer: ```suggestion
Quality Flags: "NIGHTTIME, USER FLAGGED"
```

Discard Before Resample: False
Resample Threshold Percentage: 75%
```

the observation data above is filtered and resampled into:
```
timestamp,value
2020-01-01 13:00:00+00:00,155.0
```
The first hour interval is dropped because 100% of timestamps in
the first hour are flagged `NIGHTTIME`. The second hour interval
is excluded because 75% of the points are  flagged with either
`NIGHTTIME`, or `USER FLAGGED`, which meets the resample threshold.

> #### Excluding intervals where all data is flagged

Reviewer: Maybe going in circles, but should this be something more like "Mixing filter types"?

> Consider a situation where we want to exclude resampled intervals where
> 100% of the  occurs during nighttime, is flagged by the user or
> the limits are exceeded. We would like to exclude any data points that
> are user flagged or exceed limits from resampled data.

Reviewer: ```suggestion
Consider a situation where we want to exclude resampled intervals where
100% of the interval occurs during nighttime. Further, we'd like to exclude
points that are flagged by the user or the limits are exceeded, and the
corresponding resampled intervals should be excluded only if no valid data
remains.
```

Using the quality flag filters:

```
Quality Flags: "USER FLAGGED,LIMITS EXCEEDED"
Discard Before Resample: True
Resample Threshold Percentage: 100%

Quality Flags: "NIGHTTIME"
Discard Before Resample: False
Resample Threshold Percentage: 100%
```

The 1 hour resampled data would be:
```
2020-01-01 12:00:00+00:00,130
2020-01-01 13:00:00+00:00,155
```

The first hour interval is excluded because all four points are
flagged with `NIGHTTIME`, which meets the 100% resample threshold.

> The second hour interval is included because only 75% of points
> are flagged with `USER FLAGGED` and `LIMITS EXCEEDED`. However,
> the points at `12:00`, `12:15`, and `12:30` are excluded before
> resampling the interval, resulting in an average of 130.

Reviewer: ```suggestion
In the the second hour interval the points at `12:00`, `12:15`, and `12:30` 
are `USER FLAGGED` or `LIMITS EXCEEDED`, so they are excluded before
resampling the interval, resulting in an average of 130. The resampled 
interval is retained because only 75% of points are flagged.
```

> The third hour is included and all datapoints are included when
> resampling.

Reviewer: ```suggestion
The third hour contains no flagged points, so it's also included after resampling.
```
